### ISSUE TITLE
symlink support and docstrings in filepath.FilePath.copyTo/moveTo

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3105]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)| @cyli reported|
|-|-|
|Trac ID|trac#3105|
|Type|enhancement|
|Created|2008-03-20 17:38:10Z|
|Branch|https://github.com/twisted/twisted/tree/symlinks-copy-move-3105-2|

Currently, if you copy a link to a directory, it copies the destination directory rather than the link.  From the comments on ticket 3098, this may be the desired behavior, except that the user should have the option to copy just the link.

Also, when recursively copying a directory, it will also copy the target of any symlinks in said directory.  Not sure if this is desired behavior either.  Maybe the option should be just either to follow all links, or follow no links.

Attachments:

* [filepath.copyTo.symlink.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/62f/62fe577db13fa2dfab66ac892614234f0889e433/dac3c7381210b991b7521a225ecda5b1d79b9fe3.patch) (6782 bytes) - added by cyli on 2008-03-23 22:34:25Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3105 3105
type__enhancement enhancement
reporter__cyli cyli
priority__high high
milestone__ 
branch__branches_symlinks_copy_move_3105_2 branches/symlinks-copy-move-3105-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1206034690000000 1206034690000000
changetime__1212340974000000 1212340974000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_1"></a>|@cyli commented|
|-|-|

moveTo has the same issue

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_2"></a>|@cyli removed owner|
|-|-|

Added followLinks, which determines whether symlinks should be copied over in `copyTo` as symlinks or as files and directories.  (`moveTo` depends upon `copyTo`, so same applies).

As it turns out `copyTo` has a lot of behavior which I'm not sure I should change, if people depend upon it behaving as it currently is.  So I just documented these peculiarities in the docstring.

For instance:
  * if destination is a symlink, `copyTo` copies the `FilePath` to the symlink's target
  * if the `FilePath` is a file and destination is a directory, it will raise an `IOError` instead of copying the `FilePath` into that directory (which is what happens when one executes cp)
  * permissions and ownership of the `FilePath` is not copied over
  * copyTo copies only the children of the `FilePath`, rather than the `FilePath` itself (I originally thought it behaved like 'cp')

As I'm trying to add symlink support to `filepath.FilePath`, the first property bothers me. I am not sure if followLinks should apply to the destination as well (if destination is a symlink).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve commented|
|-|-|

(In [23066]) Branching to 'symlinks-copy-move-3105'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

(In [23102]) Add more tests for moveTo.

Refs [#3105](https://github.com/twisted/twisted/issues/3105).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve commented|
|-|-|

I got my hands dirty by adding some tests and making some cleanups. This is up for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @therve|
|-|-|

Some minor issues:

  * Can you make the tests use `assert*` instead of `fail*`?
  * `test_moveTo` misspells _with_ in its docstring
  * The other `FilePath` branch has a helper `symlink` which can be used instead of `os.symlink` and which will skip the test if the platform is missing symlink support.  Maybe this branch can be merged forward after that is merged and use that instead of doing the explicit `.skip` attribute thing.
  * Docstrings for `copyTo` and `moveTo` are great.  A couple minor points about them:
    * usage of ''it'' to refer to either the function or the instance is somewhat awkward.  I'd replace uses where it refers to the function with ''this method''.
    * (`copyTo`) related, _itself_ just seems to confuse the matter.  stick with _it_ to refer to the instance.
    * (`copyTo`) _otherwords_ is two words
    * indent 2nd line of `destination` and `followLinks` `@param` markup

More significantly, the parenthetical in the docstring of `moveTo` sounds more like a bug description than desirable behavior. :/   Overall, I wonder if `copyTo` and `moveTo` aren't trying to something that's too complex without having enough information.  Their behaviors seem idiosyncratic at best(wrong at worst).  Documented the bug is still a minor improvement to the code, I suppose.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

(In [23697]) Some review comments.

Refs [#3105](https://github.com/twisted/twisted/issues/3105)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve commented|
|-|-|

(In [23698]) Remove wrong argument.

Refs [#3105](https://github.com/twisted/twisted/issues/3105)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_9"></a>|@therve commented|
|-|-|

(In [23720]) Branching to 'symlinks-copy-move-3105-2'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_10"></a>|@therve removed owner|
|-|-|

OK everything is done.

Regarding your last point, I don't think there was a problem, so removed the message. Maybe I'm wrong, though.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @therve|
|-|-|

`copyTo` should use `linkTo` instead of `os.symlink`, now that this is a possibility.

I thought there was `FilePath.readlink` too, but maybe that's still sitting off in a branch someplace.

Code looks good, as far as it goes.  I'm unhappy with both `copyTo` and `moveTo` now, though.  I wonder if I should try to make something better.  Maybe not.

Tests pass on all supported builders, I guess this should be merged.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve set status to `closed`|
|-|-|

(In [23810]) Merge symlinks-copy-move-3105-2

Authors: cyli, therve
Reviewer: exarkun
Fixes [#3105](https://github.com/twisted/twisted/issues/3105)

Add support for symlink in FilePath.copyTo and File.path.moveTo: allow to
disable link following.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
